### PR TITLE
go/seed: implement missing consensus methods

### PR DESCRIPTION
**Pull request by ptrus**
Fixes: https://github.com/oasisprotocol/oasis-core/issues/3356

Seed now supports `oasis-node control` methods and shouldn't panic on any cli requests.

TODO:
- [x] consider adding an e2e test for seed control cli

### CODE REVIEW

File: go/consensus/tendermint/seed/seed.go [hunk 1]
```diff
@@ -83,6 +95,106 @@ func (srv *seedService) Cleanup() {
...
+	for _, tmpeer := range tmpeers {
+		p := string(tmpeer.ID()) + "@" + tmpeer.RemoteAddr().String()
+		peers = append(peers, p)
+	}
+	status.NodePeers = peers
+
+	return status, nil
+}
+
+// Unimplemented consensus Backend methods.
```
Comment (ptrus) [inline go/consensus/tendermint/seed/seed.go:125]: The only missing methods are now the backends (epochtime, registry, beacon...) and `ConsensusKey` (which is only used in tests anyway). 

If we want, we could implement those remaining methods here (can still just panic on those), and get rid of the `consensus/api.BaseBackend` since it's not used anywhere else.

cc @kostko

File: go/consensus/tendermint/seed/seed.go [hunk 2]
```diff
@@ -10,24 +11,35 @@ import (
...
 	genesis "github.com/oasisprotocol/oasis-core/go/genesis/api"
 	registry "github.com/oasisprotocol/oasis-core/go/registry/api"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/syncer"
 )
 
-var logger = logging.GetLogger("consensus/tendermint/seed")
+var (
+	logger = logging.GetLogger("consensus/tendermint/seed")
+
+	errNotSupported = fmt.Errorf("not supported by seed node")
```
Comment (kostko) [inline go/consensus/tendermint/seed/seed.go:35]: Consider using `consensus.ErrUnsupported` instead.

File: go/consensus/tendermint/seed/seed.go (hunk 1 above)
Comment (kostko) [reply go/consensus/tendermint/seed/seed.go:125]: Agreed, implement the methods and remove the base backend.

**Review by kostko — COMMENTED** [review]
Cool, some minor nits and yeah a test would be useful (and should be simple to do).

File: go/consensus/tendermint/common/common.go [hunk 3]
```diff
@@ -32,11 +40,60 @@ const (
...
+	// CfgP2PSendRate is the rate at which packets can be sent, in bytes/second.
+	CfgP2PSendRate = "consensus.tendermint.p2p.send_rate"
+	// CfgP2PRecvRate is the rate at which packets can be received, in bytes/second.
+	CfgP2PRecvRate = "consensus.tendermint.p2p.recv_rate"
 )
 
 // Flags has the configuration flags.
 var Flags = flag.NewFlagSet("", flag.ContinueOnError)
 
+// GetAddress returns the configured tendermint address.
```
Comment (kostko) [inline go/consensus/tendermint/common/common.go:59]: Maybe name something like `GetExternalAddress` to be more clear.

File: go/consensus/tendermint/seed/seed.go [hunk 4]
```diff
@@ -83,6 +99,192 @@ func (srv *seedService) Cleanup() {
 	// No cleanup in particular.
 }
 
+// Synced returns a channel that is closed once synchronization is complete.
```
Comment (kostko) [inline go/consensus/tendermint/seed/seed.go:102]:
```suggestion
// Implements Backend.
```

File: go/consensus/tendermint/seed/seed.go [hunk 5]
```diff
@@ -83,6 +99,192 @@ func (srv *seedService) Cleanup() {
...
+func (srv *seedService) Synced() <-chan struct{} {
+	// Seed is always considered synced.
+	ch := make(chan struct{})
+	close(ch)
+	return ch
+}
+
+// SupportedFeatures returns the features supported by this consensus backend.
```
Comment (kostko) [inline go/consensus/tendermint/seed/seed.go:110]:
```suggestion
// Implements Backend.
```

File: go/consensus/tendermint/seed/seed.go [hunk 6]
```diff
@@ -83,6 +99,192 @@ func (srv *seedService) Cleanup() {
...
+		Backend:          api.BackendName,
+		Features:         srv.SupportedFeatures(),
+	}
+
+	// List of consensus peers.
+	tmpeers := srv.p2pSwitch.Peers().List()
+	peers := make([]string, 0, len(tmpeers))
+	for _, tmpeer := range tmpeers {
+		p := string(tmpeer.ID()) + "@" + tmpeer.RemoteAddr().String()
+		peers = append(peers, p)
```
Comment (kostko) [inline go/consensus/tendermint/seed/seed.go:128]: No peers in E2E tests when querying seed node status (according to Codecov at least)?

File: go/consensus/tendermint/seed/seed.go [hunk 7]
```diff
@@ -92,49 +294,56 @@ func New(dataDir string, identity *identity.Identity, genesisProvider genesis.Pr
...
+	srv.doc = doc
 
 	nodeInfo := p2p.DefaultNodeInfo{
 		ProtocolVersion: p2p.NewProtocolVersion(
-			version.P2PProtocol,
-			version.BlockProtocol,
-			0,
+			tmversion.P2PProtocol,
+			tmversion.BlockProtocol,
+			version.ConsensusProtocol.ToU64(),
```
Comment (kostko) [inline go/consensus/tendermint/seed/seed.go:329]: Nice catch! @tjanez don't forget about this one in #3360 (depends on which PR gets merged first).

File: go/oasis-node/cmd/node/node.go [hunk 8]
```diff
@@ -526,7 +525,8 @@ func NewTestNode() (*Node, error) {
 	return newNode(true)
 }
 
-func newNode(testNode bool) (n *Node, err error) { // nolint: gocyclo
+func newNode(testNode bool) (*Node, error) { // nolint: gocyclo
+	var err error
```
Comment (kostko) [inline go/oasis-node/cmd/node/node.go:529]: The reason for having that named `err` return value is for the deferred func below to propagate the error. This prevents that from happening. Maybe add a comment so this doesn't get changed again :-)

Comment (ptrus) [reply go/oasis-node/cmd/node/node.go:529]: Oh right, however in that case any objections to doing `(node *Node, err error)` ? (and not using a separate `n` variable for node here). Because otherwise having `n` in scope through the function which is always `nil` is a bit dangerous :grin: that's also what lead me to removing the named returns here. The only "downside" will be that a non-nil `node`  will be returned in case this function errors, but i think that's fine as?

File: go/consensus/tendermint/seed/seed.go (hunk 6 above)
Comment (ptrus) [reply go/consensus/tendermint/seed/seed.go:128]: Yeah i looked into this earlier today, so there's two things at work here:
- for inbound peers - seed instantly returns the peers and disconnects (as expected)
- for outbound - since we don't set the `"consensus.tendermint.core.external_address"` flag in the tests, nodes connecting to the seed provide `0.0.0.0:<port>` as their external address, the seed then doesn't insert the nodes to its address book and therefore also never makes outbound connections to those

Actually with current test setup seed will only ever return the peers from the genesis and never includes new nodes (but this is enough for all the tests). I also tested this by disabling the `populateAddrBookFromGenesis` in the seed and in that case none of the tests pass. However then configuring the test nodes to provide the correct external addresses (`127.0.0.1:<port>`), things work again correctly :-) - since this is working as expected i didn't commit the test here. However thinking about it more, it might make sense including such a test (disabling automatically including genesis peers in the seeds addressbook) in the suite so that we properly test seed including new peers in its address book.

Comment (ptrus) [reply go/consensus/tendermint/seed/seed.go:134]: Fixed by configuring `external_addresses` in tests so seed does outbound requests now, also a check is done in the seed test.

File: go/oasis-node/cmd/node/node.go (hunk 8 above)
Comment (kostko) [reply go/oasis-node/cmd/node/node.go:529]: So we did have it like that but changed it in d0551abb74b878bf62b44115fd917a9786c9357b due to a "possible nil dereference", but can't currently think of a case where this would be problematic.

Comment (ptrus) [reply go/oasis-node/cmd/node/node.go]: hm, so either whatever triggered that was fixed already (or it was a [false positive](https://github.com/dominikh/go-tools/issues/641)) since it doesn't seem to happen anymore

File: go/consensus/tendermint/seed/seed.go (hunk 6 above)
Comment (ptrus) [reply go/consensus/tendermint/seed/seed.go:143]: Also added support for configuring multiple seeds in the e2e tests, and a debug flag for disabling "populating address from genesis", so seed returning new peers is tested as well. Added an E2E test that tests both.

**Review by kostko — APPROVED** [review]

File: go/oasis-node/cmd/node/node.go (hunk 8 above)
Comment (ptrus) [reply go/oasis-node/cmd/node/node.go]: ah i did found it, so `node` was `nil` in the deferred function. Fixed it by passing `node` into the deferred function via an argument
